### ISSUE TITLE
Adding Regularizer in Cross Entropy Loss for Repeated Words

### ISSUE THREAD

**Issue by NickShahML** [tensorflow/tensor2tensor#369]
Hey T2T team,

Great repo. I've been trying to investigate how to add a penalty for repeated words which occur sometimes in translations. For example "He went to get milk milk milk". 

Inspired by coverage, I think the best way to do this is to add the following regulization term:

```
# For all timesteps, we add the softmaxed logits. This tells the total probability of particular words guessed.
logits_summed_across_timesteps = tf.reduce_sum(
    [tf.nn.softmax(logits) for logits in all_timesteps)

# If there are any words guessed more than once (greater probability than 1.0), then we add this to the loss
regularization = tf.reduce_sum(
    tf.maximum(1.0,logits_summed_across_timesteps)  - 1.0))
```

How can this be done for the transformer? In the `padded_cross_entropy` function, it seems that you can only see _one_ timestep. So I'm not sure how you can implement this for all timesteps. Also, this would require the softmaxed logits. Since `softmax_cross_entropy_with_logits` does the softmax internally in tensorflow, you may have to compute the softmax a separate time?

Any help with this would be greatly appreciated!

**Comment by lucasgreene** [comment]
I'm working on a similar bit. Not sure why the model is repeating itself so much. In any case I think the way you want to go about it is to operate on the logits, not the output of the softmax. The general idea that Ive implemented is for the output logit at each time step, compare to all outputs in and previous timesteps (essentially the same operation as self-attention). The comparison here will be the cosine simularity between the logits. This will be high when the model is repeating itself and low otherwise. Cosine is between [-1,1]. Ill make the simplifying assumption that we can just take the maximum reduction over time of these cosine values as our loss. You could also do a mean reduction or a sum reduction. So now we have a single cosine similarity score for each timestep (which is the maximum similarity of that output with all previous outputs). We will threshold those values so everything below some value (say 0.8) gets masked to 0. Now we have a max cosine score if its above 0.8 and 0 else. We can use incorporate these scores into the loss function by multiplying them by some hyperparameter (scaling) and then adding to the regular loss for each timestep. The only additional thing I think is important is controlling the gradient flow for the similarity loss so that it only propagates to the output that is farther forward in time. That way we dont penalize the model for outputting something that gets repeated later, only for repeating something that has already been output.  Not sure where you would put this loss in the t2t framework as I use it mostly as a library and pull out the various pieces i need

**Comment by lucasgreene** [comment]
Here is a blog explaining a similar but slightly different approach http://www.abigailsee.com/2017/04/16/taming-rnns-for-better-summarization.html . They keep track of the produced attention states and penalize the model for attending to the same thing multiple times. I feel like penalizing the model for producing similar outputs is more general, as the model could repeat itself without using attention. In any case they report success with this approach and I have yet to make the cosine loss reduce the repetition.

**Comment by NickShahML** [comment]
Thanks @lucasgreene. I talked to Abigail a while back and I definitely their coverage approach is promising. However, with multiple attention heads, I'm not totally sure how this technique can be applied. Applying this to each head would occur a huge amount of memory usage, and performance would probably drop. 

This is why I advocated penalizing by adding an extra penalty for words that have been already used in the softmax approach. I actually think this is the perferred method given what has been happening with MCTS in alpha go zero.

**Comment by rsepassi** [comment]
If you implement your own Modality, you can override the `loss` function.

**Comment by wenkesj** [comment]
@rsepassi how do you use an implemented `Modality` using `t2t-trainer`? I understand how to _register_ them, as in:

```python
@registry.register_symbol_modality("foo")
class FooSymbolModality(SymbolModality):
  """SymbolModality that uses Foo loss."""

  def loss(self, logits, targets):
    # ...
```

But not sure how to tap into the registry

**Comment by martinpopel** [comment]
@wenkesj: I think you are looking for `--t2t_usr_dir`, see https://github.com/tensorflow/tensor2tensor#adding-your-own-components

**Comment by wenkesj** [comment]
@martinpopel Yes, I understand how to register them.  I want to know how to target the use of the _specific_ modality on the problem/model, i.e. replacing the `CTCSymbolModality` with `FooSymbolModality`

**Comment by martinpopel** [comment]
Oh, sorry, I did not understand. I have never played with the modalities, but checking the code I see https://github.com/tensorflow/tensor2tensor/blob/a1d7ed7e7d96ea0d89ca153ba46b7f51d394580f/tensor2tensor/layers/common_hparams.py#L136-L143

**Comment by wenkesj** [comment]
@martinpopel NICE! Thanks

**Comment by Chris19920210** [comment]
@NickShahML Have you ever added the penalty for repeated works? Does that work?  I'm facing the same problem now. Thx.
